src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs

```diff
@@ -1142,6 +1191,53 @@ private static AckMessage CreateAckMessage(ref MessagePackReader reader, int arr
             return result;
         }
 
+        private static ClientInvocationMessage CreateClientInvocationMessage(ref MessagePackReader reader, int arrayLength)
+        {
+            var invocationId = ReadString(ref reader, "invocationId");
+            var connectionId = ReadString(ref reader, "connectionId");
+            var callerServerId = ReadString(ref reader, "callerServerId");
+            var payloads = ReadPayloads(ref reader);
+
+            var result = new ClientInvocationMessage(invocationId, connectionId, callerServerId, payloads);
+            result.ReadExtensionMembers(ref reader);
+            return result;
+        }
+
+        private static ServiceCompletionMessage CreateServiceCompletionMessage(ref MessagePackReader reader, int arrayLength)
+        {
+            var invocationId = ReadString(ref reader, "invocationId");
+            var connectionId = ReadString(ref reader, "connectionId");
+            var callerServerId = ReadString(ref reader, "callerServerId");
+            var error = ReadString(ref reader, "error");
+            var protocol = ReadString(ref reader, "protocol");
+            var payload = ReadBytes(ref reader, "payload");
+
+            ServiceCompletionMessage result;
+            if (string.IsNullOrEmpty(error))
+            {
+                result = new ServiceCompletionMessage(invocationId, connectionId, callerServerId, protocol, payload);
```

Comment:
how about another ErrorCompletionMessage?

Reply:
Any concern about putting them together? Reviewed SignalR [`CompletionMessage`](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/SignalR.Common/src/Protocol/CompletionMessage.cs#L16) as well as service protocols `HandshakeResponseMessage` and `AccessKeyResponseMessage`, looks we usually make success and failure in the same type.

Reply:
separating looks clearer to me. Now the parser relies on additional logic(always check error first) to parse the message, and when the error response contains more properties, it is mixing properties for different purposes together.

Reply:
But since we already have such convention, I am fine to follow the current convention

Reply:
Make sense. I'm persuaded and would split to abstract `ServiceCompetionMessage` with 2 classes `ClientCompletionMessage` and `ErrorCompletionMessage`.